typing/rec_check.ml

```diff
@@ -193,15 +193,26 @@ let array_element_kind env ty =
             when List.for_all (fun c -> c.Types.cd_args = Types.Cstr_tuple [])
                 cstrs ->
               `Pintarray
-          | {type_kind = _} ->
+          | {type_kind = Type_variant _ | Type_open | Type_record _} ->
               `Paddrarray
         with Not_found ->
           (* This can happen due to e.g. missing -I options,
               causing some .cmi files to be unavailable.
               Maybe we should emit a warning. *)
           `Pgenarray
       end
-  | _ ->
+  | Tlink _ | Tsubst _ ->
+      (* [Typeopt.scrape_ty] starts with [Ctype.expand_head_opt],
+         so these cases are impossible *)
+      assert false
+  | Tpoly(t, _alphas) ->
+      (* this case should not occur after [scrape_ty],
+         but we handle it defensively *)
+      array_element_kind env t
```

Comment:
Again, I don't understand why. Seems like `repr (expand_head_opt (Tpoly(_))` would just be a noop. What am I missing?